### ISSUE TITLE
Optimizer + tf.reduce_prod() + GPU = Cannot assign a device to node 'gradients/Prod_grad/ListDiff'

### ISSUE THREAD

**Issue by bsautermeister** [tensorflow/tensorflow#3957]
### Environment info
Operating System: _Ubuntu 14.4.5 LTS_

Installed version of CUDA and cuDNN: 
_CUDA 7.5.18_ and _cuDNN 4.0.7_
```
-rw-r--r-- 1 root root 322936 Aug 15  2015 libcudadevrt.a
lrwxrwxrwx 1 root root     16 Jul 12 08:24 libcudart.so -> libcudart.so.7.5
lrwxrwxrwx 1 root root     19 Jul 12 08:24 libcudart.so.7.5 -> libcudart.so.7.5.18
-rwxr-xr-x 1 root root 383336 Aug 15  2015 libcudart.so.7.5.18
-rw-r--r-- 1 root root 720192 Aug 15  2015 libcudart_static.a

-rw-r--r-- 1 sauterme tumuser        0 Aug 22 18:29 libcudart.so.7.5
-rw-r--r-- 1 sauterme tumuser        0 Aug 22 18:29 libcudart.so.7.5.18
lrwxrwxrwx 1 sauterme tumuser       13 Feb  9  2016 libcudnn.so -> libcudnn.so.4
lrwxrwxrwx 1 sauterme tumuser       17 Feb  9  2016 libcudnn.so.4 -> libcudnn.so.4.0.7
-rwxr-xr-x 1 sauterme tumuser 61453024 Feb  8  2016 libcudnn.so.4.0.7
-rw-r--r-- 1 sauterme tumuser 62025862 Feb  8  2016 libcudnn_static.a
```

If installed from binary pip package, provide:

1. Which pip package you installed:
pip package version 0.10 for Python 2.7 with GPU support:
_https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow-0.10.0rc0-cp27-none-linux_x86_64.whl_

2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.
_0.10.0rc0_

### Steps to reproduce
1. Use a system with GPU
2. Use this code snippet

  ```
  import tensorflow as tf
  
  g = tf.Graph()
  with g.as_default():
      with tf.Session(config=tf.ConfigProto(log_device_placement=True)) as sess:
  
          with tf.device('/gpu:0'):
  
              a = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[2, 3], name='a')
              b = tf.Variable([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0], [7.0, 8.0, 9.0]], name='b')
  
              c = tf.matmul(a, b)
              res = tf.reduce_prod(c)
  
              #with tf.device('/cpu:0'): # <<< adding this solves the issue!
              train_op = tf.train.AdagradOptimizer(1e-3).minimize(res)
  
          sess.run(tf.initialize_all_variables())    
  
          print sess.run(res)
          print sess.run(train_op)
  
  ```
3. You will get an InvalidArgumentError:
_InvalidArgumentError: Cannot assign a device to node 'gradients/Prod_grad/ListDiff': Could not satisfy explicit device specification '/device:GPU:0' because no supported kernel for GPU devices is available.
	 [[Node: gradients/Prod_grad/ListDiff = ListDiff[T=DT_INT32, _device="/device:GPU:0"](gradients/Prod_grad/range_1, Const)]]
Caused by op u'gradients/Prod_grad/ListDiff', defined at:
  File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals_ 

4. Cause of the issue: tf.reduce_prod() when running the train_op. So, the problem takes place while back-propagation through this op.

### What have you tried?
1. Wrapping the tf.reduce_prod() with tf.device('/cpu:0'):

  Result: still getting this error!

2. Exchangeing tf.reduce_prod() with other reduce_*() ops:

  Result: no error!

3. Wrapping the Optimizer with  tf.device('/cpu:0'):

  **Result: Solves the error!**

**Comment by prb12** [comment]
@ibab @girving Possibly related to #3351 ?

**Comment by girving** [comment]
@vrv We were discussing a similar thing recently.  Is the right solution here just to tell people to leave the optimizer out of a location?  That seems heavy handed, but I don't think we can guarantee that gradients use no CPU-only ops.

**Comment by vrv** [comment]
Well, the answer that I personally don't like is to tell people to use soft_placement in the config options.  A better solution would be to implement the GPU versions of these ops when possible, and if not, to take the ops that are not implemented and wrap them in a with device(cpu:0) block.

**Comment by vrv** [comment]
(the latter approach is an implementation detail of the gradient that we can change).

**Comment by girving** [comment]
@bsautermeister Does this commit fix your problem?  https://github.com/girving/tensorflow/commit/957fe73546f25c7f88232ed560ed285c1c97067d

**Comment by bsautermeister** [comment]
@girving Indeed, it does! Thank you very much!

And (obviously) it's much faster than wrapping the Optimizer with _tf.device('/cpu:0')_.

**Comment by girving** [comment]
We shouldn't close it  until I actually submit the fix.

**Comment by carusyte** [comment]
My model's performance degraded seriously because of this. tf version is 1.9, hoping the fix could be considered officially
